build_tools/cmake/iree_bytecode_module.cmake

```diff
@@ -92,6 +104,31 @@ function(iree_bytecode_module)
     get_filename_component(_FRIENDLY_NAME "${_RULE_SRC}" NAME)
   endif()
 
+  # Enforce that IREE_ENABLE_TSAN agrees with IREE_BYTECODE_MODULE_ENABLE_TSAN,
+  # but only if iree_bytecode_module is called.
+  #
+  # That way, we allow people who don't build iree_bytecode_module's to flip
+  # only IREE_ENABLE_TSAN and not have to care about IREE_BYTECODE_MODULE_*.
+  #
+  # STREQUAL feels wrong here - we don't care about the exact true-value used,
+  # ON or TRUE or something else. But we haven't been able to think of a less bad
+  # alternative. https://github.com/google/iree/pull/8474#discussion_r840790062
+  if(NOT IREE_ENABLE_TSAN STREQUAL IREE_BYTECODE_MODULE_ENABLE_TSAN)
+    list(APPEND _CUSTOM_BUILD_TIME_CMAKE_ERROR_TARGETS
+      error_on_mismatched_tsan_options)
```

Comment:
I don't see enough reason to move this check to build time from configure time. If you configure with options that aren't compatible, that should be an error _at configure time_ - not an error when _building_ specific targets.

We could put the check after
```cmake
  if(_RULE_TESTONLY AND NOT IREE_BUILD_TESTS)
    return()
  endif()
```
(still in this file), to run at configure time _if_ any `iree_bytecode_module` targets are actually generated.

---

If keeping this at build time, I think we'd also want to exclude the custom target from `ALL`, e.g.
```cmake
set_property(TARGET ${_NAME} PROPERTY EXCLUDE_FROM_ALL ON)
```
(with `error_on_mismatched_tsan_options`)

Reply:
The rationale to push this past configure time is that at configure time we only have very coarse information (few booleans like `IREE_BUILD_TESTS`) about whether we will be building bytecode modules.

The concrete scenario is:
1. Dev is working with a build dir that as `IREE_BUILD_TESTS=ON`.
2. Dev wants to rerun a particular program with TSan - not involving rebuilding a `iree_bytecode_module`.
3. Dev flips `IREE_ENABLE_TSAN` and rebuilds. Does it work out of the box?

Reply:
Oops, missed this reply (without rerequesting review a PR will drop out of my queue).

> at configure time we only have very coarse information (few booleans like IREE_BUILD_TESTS) about whether we will be building bytecode modules

My view is that configure time should be checking for as much as possible. If the project configures successfully, you should be able to build `all`, `install`, and all the other targets that CMake makes available. I'm opposed to anything that blurs that line. Fail fast, and with a clear error message.

Reply:
While I appreciate the intent of this PR, I think I agree with Scott. I think we should instead enable people to not configure `iree_bytecode_module`s at all. So just like if you don't care about the compiler you can disable it, if you don't care about creating IREE vmfb's, you can disable that as well.